pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/table/internal/SymTableFactory.java

```diff
@@ -386,15 +391,26 @@ JSymbolTable recordCtor(JSymbolTable parent, JClassType recordType, JConstructor
     JSymbolTable localVarSymTable(JSymbolTable parent, JClassType enclosing, Iterable<ASTVariableId> ids) {
         List<JVariableSig> sigs = new ArrayList<>();
         for (ASTVariableId id : ids) {
+            if (unnamedVariableIsSupported() && id.isUnnamed()) {
+                continue;
+            }
             sigs.add(id.getTypeSystem().sigOf(enclosing, (JLocalVariableSymbol) id.getSymbol()));
         }
         return SymbolTableImpl.withVars(parent, VARS.augment(varNode(parent), false, ScopeInfo.LOCAL, VARS.groupByName(sigs)));
     }
 
     JSymbolTable localVarSymTable(JSymbolTable parent, JClassType enclosing, JVariableSymbol id) {
+        if (unnamedVariableIsSupported() && "_".equals(id.getSimpleName())) {
+            // unnamed variable
+            return parent;
+        }
         return SymbolTableImpl.withVars(parent, VARS.augment(varNode(parent), false, ScopeInfo.LOCAL, id.getTypeSystem().sigOf(enclosing, (JLocalVariableSymbol) id)));
     }
 
+    private boolean unnamedVariableIsSupported() {
+        return processor.getJdkVersion() >= 22;
+    }
+
```

Comment:
unnamed variables are actually a preview feature in Java 21, so this check is incorrect…

Having said that, I'd rather just remove the check, and, as we do in other places, simply assume the code compiles / let the parser do the validations

Reply:
And we are actually already checking it during parsing… https://github.com/pmd/pmd/blob/feddca0f4a6e839d5856358060067ab1745f7862/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/internal/LanguageLevelChecker.java#L139

Reply:
Ok, I think I understand why you did this…

So, technically, code written for Java 1.3 through 8 (among the versions supported by PMD), could use `_` as an identifier, without it meaning "unnamed var". Java 9+ would fail to compile (it's a keyword), and 21-preview + would recognize it as unnamed…

So, I think we have 3 alternatives:
 * we ignore this. Remove the check and have it treat "_" as unnamed for Java 1.3 though 8 (assuming there is very little code if any that used this). Notice that `ASTVariableId.isUnnamed` behaves this way in this PR, it will say true regardless of the Java version, this PR only filters in the sym table.
 * we have `LanguageLevelChecker` mark the `ASTVariableId` as unnamed only for supported versions, and we use that directly all-over (would be consistent and compliant with the Java Language in all versions, but having `LanguageLevelChecker` modify the AST is weird… maybe there is somewhere else we can do this?)
 * we expose the `LanguageLevelChecker` check so it can be consumed for specific checks directly from other places (ie: `ASTVariableId`). Not a big fan of this one…

Thoughts?

Reply:
Preview versions are not meant to stay so I thought we could just ignore the java 21 preview and not need to update this when we remove it... I don't think anyone will notice

About ASTVariableDeclaratorId::isUnnamed, i would have it just check the version internally. No need to change anything to LanguageLevelChecker